URL validation for multiple hash signs in URL

RFC considers URLs with multiple hashes as invalid. However, browsers will try their best to read non-valid URLs by treating repeat hashes as though they are escaped.

STR:
- Load `https://playground.babylonjs.com/debug.html#NT2UCM#13`
- Open a new window: the title bar of the babylon URL is empty
- There are a bunch of telemetry errors related to URI parse:
```
2020-07-02 15:50:47.439 26419-26419/org.mozilla.vrbrowser.dev E/VRB[GleanMetricsService]: Invalid URL
    java.lang.IllegalArgumentException: Illegal character in fragment at index 50: https://playground.babylonjs.com/debug.html#NT2UCM#13
        at java.net.URI.create(URI.java:854)
        at org.mozilla.vrbrowser.telemetry.GleanMetricsService.stopPageLoadTimeWithURI(GleanMetricsService.java:111)
        at org.mozilla.vrbrowser.browser.engine.Session.onPageStop(Session.java:1187)
        at org.mozilla.geckoview.GeckoSession$5.handleMessage(GeckoSession.java:622)
        at org.mozilla.geckoview.GeckoSession$5.handleMessage(GeckoSession.java:611)
        at org.mozilla.geckoview.GeckoSessionHandler.handleMessage(GeckoSessionHandler.java:90)
        at org.mozilla.gecko.EventDispatcher$3.run(EventDispatcher.java:356)
        at android.os.Handler.handleCallback(Handler.java:751)
        at android.os.Handler.dispatchMessage(Handler.java:95)
        at android.os.Looper.loop(Looper.java:154)
        at android.app.ActivityThread.main(ActivityThread.java:6165)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
     Caused by: java.net.URISyntaxException: Illegal character in fragment at index 50: https://playground.babylonjs.com/debug.html#NT2UCM#13
        at java.net.URI$Parser.fail(URI.java:2856)
        at java.net.URI$Parser.checkChars(URI.java:3029)
        at java.net.URI$Parser.parse(URI.java:3075)
        at java.net.URI.<init>(URI.java:590)
        at java.net.URI.create(URI.java:852)
        at org.mozilla.vrbrowser.telemetry.GleanMetricsService.stopPageLoadTimeWithURI(GleanMetricsService.java:111) 
        at org.mozilla.vrbrowser.browser.engine.Session.onPageStop(Session.java:1187) 
        at org.mozilla.geckoview.GeckoSession$5.handleMessage(GeckoSession.java:622) 
        at org.mozilla.geckoview.GeckoSession$5.handleMessage(GeckoSession.java:611) 
        at org.mozilla.geckoview.GeckoSessionHandler.handleMessage(GeckoSessionHandler.java:90) 
        at org.mozilla.gecko.EventDispatcher$3.run(EventDispatcher.java:356) 
        at android.os.Handler.handleCallback(Handler.java:751) 
        at android.os.Handler.dispatchMessage(Handler.java:95) 
        at android.os.Looper.loop(Looper.java:154) 
        at android.app.ActivityThread.main(ActivityThread.java:6165) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776) 
```

--- a/app/src/common/shared/org/mozilla/vrbrowser/telemetry/GleanMetricsService.java
+++ b/app/src/common/shared/org/mozilla/vrbrowser/telemetry/GleanMetricsService.java
@@ -26,6 +26,7 @@ import org.mozilla.vrbrowser.utils.SystemUtils;
 import org.mozilla.vrbrowser.utils.UrlUtils;
 
 import java.net.URI;
+import java.net.URISyntaxException;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Hashtable;
@@ -108,7 +109,7 @@ public class GleanMetricsService {
         }
 
         try {
-            URI uriLink = URI.create(uri);
+            URI uriLink = UrlUtils.parseUri(uri);
             if (uriLink.getHost() == null) {
                 return;
             }
@@ -117,7 +118,7 @@ public class GleanMetricsService {
                 Url.INSTANCE.domains().add();
             }
             Url.INSTANCE.visits().add();
-        } catch (IllegalArgumentException e) {
+        } catch (URISyntaxException e) {
             Log.e(LOGTAG, "Invalid URL", e);
         }
     }

--- a/app/src/common/shared/org/mozilla/vrbrowser/telemetry/TelemetryWrapper.java
+++ b/app/src/common/shared/org/mozilla/vrbrowser/telemetry/TelemetryWrapper.java
@@ -31,6 +31,7 @@ import org.mozilla.vrbrowser.utils.SystemUtils;
 import org.mozilla.vrbrowser.utils.UrlUtils;
 
 import java.net.URI;
+import java.net.URISyntaxException;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Map;
@@ -307,7 +308,7 @@ public class TelemetryWrapper {
         }
 
         try {
-            URI uriLink = URI.create(uri);
+            URI uriLink = UrlUtils.parseUri(uri);
             if (uriLink.getHost() == null) {
                 return;
             }
@@ -332,7 +333,7 @@ public class TelemetryWrapper {
 
             loadingTimeHistogram[histogramLoadIndex]++;
 
-        } catch (IllegalArgumentException e) {
+        } catch (URISyntaxException e) {
             Log.e(LOGTAG, "Invalid URL", e);
         }
     }

--- a/app/src/common/shared/org/mozilla/vrbrowser/utils/StringUtils.java
+++ b/app/src/common/shared/org/mozilla/vrbrowser/utils/StringUtils.java
@@ -76,6 +76,10 @@ public class StringUtils {
         return false;
     }
 
+    public static long charCount(@NonNull String aString, char target) {
+        return aString.chars().filter(ch -> ch == target).count();
+    }
+
     /**
      * The version code is composed like: yDDDHHmm
      *  * y   = Double digit year, with 16 substracted: 2017 -> 17 -> 1

--- a/app/src/common/shared/org/mozilla/vrbrowser/utils/UrlUtils.java
+++ b/app/src/common/shared/org/mozilla/vrbrowser/utils/UrlUtils.java
@@ -22,6 +22,7 @@ import org.mozilla.vrbrowser.telemetry.TelemetryWrapper;
 import java.io.File;
 import java.net.MalformedURLException;
 import java.net.URI;
+import java.net.URISyntaxException;
 import java.net.URL;
 import java.util.regex.Pattern;
 
@@ -139,14 +140,14 @@ public class UrlUtils {
 
             } else {
                 try {
-                    URI uri = URI.create(aUri);
+                    URI uri = parseUri(aUri);
                     URL url = new URL(
                             uri.getScheme() != null ? uri.getScheme() : "",
                             uri.getAuthority() != null ? uri.getAuthority() : "",
                             "");
                     return url.toString();
 
-                } catch (MalformedURLException | IllegalArgumentException e) {
+                } catch (MalformedURLException | URISyntaxException e) {
                     return "";
                 }
             }
@@ -199,6 +200,21 @@ public class UrlUtils {
             return url.getHost();
         } catch (MalformedURLException e) {
             return uri;
+        }
+    }
+
+    public static URI parseUri(String aUri) throws URISyntaxException {
+        try {
+            return new URI(aUri);
+        } catch (URISyntaxException e) {
+            if (!StringUtils.isEmpty(aUri) && StringUtils.charCount(aUri, '#') >= 2) {
+                // Browsers are able to handle URLs with double # by ignoring everything after the
+                // second # occurrence. But Java implementation considers it an invalid URL.
+                // Remove everything after the second #.
+                int index = aUri.indexOf("#", aUri.indexOf("#") + 1);
+                return parseUri(aUri.substring(0, index));
+            }
+            throw e;
         }
     }
 

